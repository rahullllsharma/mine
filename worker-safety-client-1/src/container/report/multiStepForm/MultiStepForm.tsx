import type { Dispatch, PropsWithChildren } from "react";
import type {
  MultiStepFormActions,
  MultiStepFormState,
  MultiStepFormStep,
} from "./state/reducer";
import { useMemo, useReducer } from "react";

import MultiStepFormContext from "./context/MultiStepFormContext";
import multiStepFormReducer from "./state/reducer";
import MultiStep from "./components/multiStep/MultiStep";
import { formatSteps } from "./utils";

export type MultiStepFormProps<T = Record<string, unknown>> = {
  steps: MultiStepFormStep<T>[];
  onStepMount?: () => any;
  onStepUnmount?: (data: unknown) => void;
  onStepChange?: (data: unknown) => void;
  onStepSave?: (
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    data: any
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ) => void | Promise<{ data?: any; errors?: any }> | Promise<boolean>;
  onComplete?: (data: unknown) => void;
  UNSAFE_WILL_BE_REMOVED_debugMode?: boolean;
  readOnly?: boolean;
};

export function MultiStepFormProvider({
  steps,
  children,
}: PropsWithChildren<Pick<MultiStepFormProps, "steps">>): JSX.Element {
  const [reducer, dispatch] = useReducer(
    multiStepFormReducer,
    steps,
    formatSteps
  );

  // stabilize internals
  const memoizedReducer = useMemo(() => [reducer, dispatch], [reducer]) as [
    MultiStepFormState,
    Dispatch<MultiStepFormActions>
  ];

  return (
    <MultiStepFormContext.Provider
      value={memoizedReducer}
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      displayName="MultiStepFormContext"
    >
      {children}
    </MultiStepFormContext.Provider>
  );
}

export default function MultiStepForm({
  steps,
  ...props
}: MultiStepFormProps): JSX.Element {
  return (
    <MultiStepFormProvider steps={steps}>
      <MultiStep {...props} />
    </MultiStepFormProvider>
  );
}
