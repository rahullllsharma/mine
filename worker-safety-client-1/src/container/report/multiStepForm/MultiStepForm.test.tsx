import type { MultiStepFormStateItem } from "./state/reducer";
import { render, screen, within } from "@testing-library/react";
import MultiStepForm from "./MultiStepForm";

jest.mock("@/hooks/useLeavePageConfirm");

jest.mock("next/router", () => ({
  useRouter: jest.fn(() => ({
    asPath: "path/to/page",
    push: jest.fn(),
  })),
}));

describe(MultiStepForm.name, () => {
  let cacheHistoryState: History;

  beforeEach(() => {
    cacheHistoryState = window.history;

    Object.defineProperty(window, "history", {
      value: {
        replaceState: jest.fn(),
        state: {
          as: "pathname",
        },
      },
      writable: true,
    });

    window.HTMLElement.prototype.scrollTo = jest.fn();
  });

  afterEach(() => {
    Object.defineProperties(
      window.history,
      cacheHistoryState as unknown as PropertyDescriptorMap
    );
  });

  it("should render the multi form properly", async () => {
    const steps = [
      {
        id: "step",
        name: "Step",
        path: "#step",
        // eslint-disable-next-line react/display-name
        Component: () => <p>Hello Step</p>,
      },
    ] as unknown as MultiStepFormStateItem[];

    const { asFragment } = render(<MultiStepForm steps={steps} />);
    await screen.findByText(/hello step/i);

    // Navigation(s)
    const nav = screen.getByTestId("multi-step-navigation");
    // non-mobile navigation
    within(nav).getByRole("tablist");
    // mobile only navigation
    within(nav).getByTestId("select");

    // Save button
    screen.getByRole("button", {
      name: /save and continue/i,
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
