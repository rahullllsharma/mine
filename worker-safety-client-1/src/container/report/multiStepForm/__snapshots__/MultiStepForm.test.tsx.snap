// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiStepForm should render the multi form properly 1`] = `
<DocumentFragment>
  <form
    class="grid gap-4 grid-rows-2-auto-expand md:grid-rows-none md:grid-cols-2-auto-expand h-full w-screen responsive-padding-x overflow-hidden pb-12"
  >
    <aside
      class="bg-white justify-self-start w-full max-w-[240px] md:overflow-y-scroll md:col-span-1 md:py-6 md:px-3 md:justify-self-stretch md:max-w-xs md:min-w-max md:w-[33vw]"
      data-multi-step-form-nav="true"
      data-testid="multi-step-navigation"
    >
      <div
        class="hidden md:block"
      >
        <div
          aria-orientation="vertical"
          class="flex flex-col gap-2"
          role="tablist"
        >
          <button
            aria-selected="true"
            class="bg-neutral-shade-3"
            id="headlessui-tabs-tab-1"
            role="tab"
            tabindex="0"
            type="button"
          >
            <div
              class="relative flex items-center px-2.5 py-3 rounded h-12 w-full hover:shadow-10 border-2 text-neutral-shade-100 cursor-pointer border-neutral-shade-100"
            >
              <i
                aria-hidden="true"
                class="text-lg mr-1 text-neutral-shade-75"
              >
                <svg
                  fill="none"
                  height="1em"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <circle
                    cx="12"
                    cy="12"
                    fill="currentColor"
                    r="10"
                  />
                </svg>
              </i>
              <span
                class="text-base mx-1 truncate"
              >
                Step
              </span>
            </div>
          </button>
        </div>
      </div>
      <div
        class="relative text-neutral-shade-100 md:hidden block"
        data-testid="select"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="w-full px-2 py-0.5 flex items-center justify-between bg-white text-base rounded border border-solid outline-none focus-within:ring-1 focus-within:ring-brand-gray-60 border-neutral-shade-26"
          id="headlessui-listbox-button-2"
          type="button"
        >
          <span
            class="truncate"
          >
            Step
          </span>
          <i
            aria-hidden="true"
            class="ci-chevron_down text-neutral-shade-58 text-xl"
          />
        </button>
      </div>
    </aside>
    <section
      class="flex flex-col gap-4 bg-white pt-6 md:pt-6 sm:pt-0 md:pb-14 overflow-auto"
    >
      <div
        class="h-full overflow-y-scroll px-6"
      >
        <p>
          Hello Step
        </p>
      </div>
      <footer
        class="flex items-center p-2 bg-brand-gray-10 absolute left-0 right-0 bottom-0"
      >
        <div
          class="flex flex-1 justify-end"
        >
          <button
            class="text-center truncate disabled:opacity-38 disabled:cursor-not-allowed text-base rounded-md font-semibold text-white bg-brand-urbint-40"
            type="button"
          >
            <div
              class="flex items-center justify-center h-full hover:bg-neutral-light-16 focus:bg-neutral-shade-7 active:bg-neutral-shade-7 px-2.5 py-2"
            >
              <span
                class="mx-1 truncate"
              >
                Save and continue
              </span>
            </div>
          </button>
        </div>
      </footer>
    </section>
  </form>
</DocumentFragment>
`;
